### DISCUSSION TITLE
Inconsistent Cookie Behavior with Axios withCredentials in Vue.js

### DISCUSSION THREAD

**Discussion by firefiber** [axios/axios#6158] [Q&A]
I'm working on a Vue.js frontend and have encountered an odd behavior regarding cookie transmission using Axios. I've set up a custom Axios instance with default configurations and an interceptor to append the CSRF token in headers. The setup is as follows:

```javascript
// Custom Axios instance
const customAxios = axios.create({
    baseURL: config.BASE,
    withCredentials: true // Global setting
});

// Axios interceptor to add CSRF token
customAxios.interceptors.request.use(function (config) {
    // Interesting things happen
    // ...
    return config;
});
``` 
I use this custom instance across various API endpoint functions, like this:

```javascript
export const userTrainingData = async () => {
    return await customAxios.get(config.BASE_USER_TRAINING, { withCredentials: true });
};
``` 

Now, I've observed the following behavior:

Global `withCredentials`: When `withCredentials` is true in the Axios instance, and I don't specify it in the endpoint function, all cookies (sessionid and csrftoken) are sent as expected.

Local `withCredentials`: When I remove the global `withCredentials` and only specify it in the individual endpoint function, only the csrftoken cookie is sent, but the sessionid cookie is missing. This occurs consistently across different browsers. I've also tested with setting withCredentials inside the interceptor itself, instead of within the create method, which results in the same behaviour (sessionid not sent, csrftoken cookie sent, along with the csrf header)

Headers are all correct (I've checked in the dev tools, access control, path, all the required ones), and the rest of the application behaves as expected. CORS/CSRF settings are all correct. It's not an issue with the cookies being sent back, or being set properly by the browser. It's just this specific discrepancy based on where withCredentials is defined that's weird. 

Could anyone explain why this might be happening? Is this expected behavior with Axios, or might there be something I'm overlooking with cookie settings or Axios configuration? Any insights or similar experiences would be greatly appreciated!

Thank you!

**Comment by jasonsaayman** [comment]
There should not be a cookie-level difference between these two Axios configs:

```js
axios.create({ withCredentials: true })
```

and:

```js
customAxios.get(url, { withCredentials: true })
```

By the time the request reaches the browser adapter, both should result in `withCredentials === true`. Axios does not choose which cookies to send; the browser does, based on the final request URL plus cookie attributes like `Domain`, `Path`, `SameSite`, `Secure`, and third-party cookie rules.

The first thing I would check is the interceptor. If it returns a new config object instead of mutating/returning the existing `config`, it may accidentally drop the per-request `withCredentials` value:

```js
customAxios.interceptors.request.use(config => {
  console.log(config.url, config.withCredentials);
  return config;
});
```

If `withCredentials` is `true` there, Axios has done its part and the difference is likely browser/cookie matching. If it is missing/false, the config is being overwritten before the request is sent.

In general, prefer setting this once on the instance if every request to that API needs cookies:

```js
const customAxios = axios.create({
  baseURL: config.BASE,
  withCredentials: true
});
```
